hackathon_site/hardware/serializers.py

```diff
@@ -46,9 +50,91 @@ def get_unique_hardware_count(obj: Category):
 
     class Meta:
         model = Order
         fields = ("id", "hardware_set", "team", "status", "created_at", "updated_at")
+
+
+class OrderCreateSerializer(serializers.Serializer):
+    class OrderCreateHardwareSerializer(serializers.Serializer):
+        id = serializers.PrimaryKeyRelatedField(
+            queryset=Hardware.objects.all(), many=False, required=True
+        )
+        quantity = serializers.IntegerField(required=True)
+
+    team_id = serializers.PrimaryKeyRelatedField(
+        queryset=TeamEvent.objects.all(), many=False, required=True
+    )
+    hardware = OrderCreateHardwareSerializer(many=True, required=True)
+
+    @staticmethod
+    def merge_requests(hardware_requests):
+        return functools.reduce(
+            lambda x, y: x + y,
+            [Counter({e["id"]: e["quantity"]}) for e in hardware_requests],
+        )
+
+    def create(self, validated_data):
+        requested_hardware = self.merge_requests(
+            hardware_requests=validated_data["hardware"]
+        )
+        relevant_past_order_items = (
+            OrderItem.objects.filter(hardware__in=requested_hardware.keys(),)
+            .exclude(order__status="Cancelled")
+            .select_related("hardware")
+        )
+        team_past_hardware_counts = Counter(
+            (
+                map(
+                    lambda item: item.hardware,
+                    relevant_past_order_items.filter(
+                        order__team=validated_data["team_id"],
+                    ).all(),
+                )
+            )
+        )
+        all_past_hardware_counts = Counter(
+            (map(lambda item: item.hardware, relevant_past_order_items.all(),))
+        )
+        new_order = None
+        unfulfilled_hardware_requests = dict()
+        for (hardware, requested_quantity) in requested_hardware.items():
+            allowed_quantity = min(
+                hardware.max_per_team - team_past_hardware_counts[hardware],
+                hardware.quantity_available - all_past_hardware_counts[hardware],
+            )
+            if allowed_quantity <= 0:
+                unfulfilled_hardware_requests[hardware] = requested_quantity
+                continue
+            if new_order is None:
+                new_order = Order.objects.create(
+                    team=validated_data["team_id"], status="Submitted"
+                )
+            if allowed_quantity < requested_quantity:
+                unfulfilled_hardware_requests[hardware] = (
+                    requested_quantity - allowed_quantity
+                )
+            for _ in itertools.repeat(None, min(allowed_quantity, requested_quantity)):
+                OrderItem.objects.create(
+                    order=new_order, hardware=hardware,
+                )
+
+        return new_order, unfulfilled_hardware_requests
+
+
+class OrderCreateResponseSerializer(serializers.Serializer):
+    def to_representation(self, instance):
```

Comment:
The downside of using `to_representation` is that we don't get proper documentation of it on swagger and such.

How annoying do you think it would be for you to create a serializer that actually uses serializer fields? We can then annotate the view with a drf_yasg decorator so that swagger knows it uses a different serializer for the response

Reply:
I didn't know about that, I'll look into it

Reply:
Here's the decorator I was talking about - [`@swagger_auto_schema`](https://drf-yasg.readthedocs.io/en/stable/custom_spec.html#the-swagger-auto-schema-decorator).

You would decorate the post method of the view something like

```
@swagger_auto_schema=(responses={201: OrderCreateResponseSerializer})
def post(...)
    ...
```

Reply:
done